Automattic/woocommerce-payments
# Ensure Subscription Initialization is only run once

Fixes #5615 

#### Changes proposed in this Pull Request


This PR addresses the issue that was discovered in #5615. In short, the issue was that subscription renewals were running multiple charges for each renewal. It was discovered that it was running transactions for as many UPE payment methods that were available + 1. We discovered this was due to the fact that the `woocommerce_scheduled_subscription_payment_woocommerce_payments` action was being added for every payment method class. This was happening since each class was ultimately inheriting from `WC_Payment_Gateway_WCPay` which calls `maybe_init_subscriptions` in it's constructor and each class was calling it's parent contructor within it's own constructor.

To address this issue we took two steps, one to explicity avoid this renewal duplicate charge issue and one to avoid any other oddities that might have been waiting in hiding due to the multiple calls to `maybe_init_subscriptions`. 

- Addressing the renewal duplicate charge issue explicitly involved checking for the existence of the `woocommerce_scheduled_subscription_payment_woocommerce_payments` before adding it. In this way we ensure that the action is only ever added once (and thus only ever called once).
- Avoiding any further oddites involved removing the `maybe_init_subscriptions` from the `WC_Payment_Gateway_WCPay` constructor and instead calling it explicitly once all the of the enabled / relevant gateways have been instantiated.







#### Testing instructions
The following test should all be performed under each of these four scenarios:
- No UPE at all.
- Legacy UPE.
- Split UPE without WooPay.
- Split UPE with WooPay.

1. Configure store for the current scenario you are testing above.
2. If not already present, install the WC Subscriptions plugin and add a subscription product.
3. Go to checkout a subscription product using a CC.
4. Confirm subscription checkout was successful and a subscription was created.
5. Now as merchant admin, go into the subscription that was created and under the subscription actions, choose "Process Renewal" and hit Update.
6. Confirm that the renewal order was created and paid under the subscription order page under "Related Orders".
7. Log into the Stripe dashboard and view the test merchant account's payments and confirm that there is only one charge for the renewal.

After confirmation of the above, run an additional test with a subscription and SEPA.
1. Enable SEPA on your merchant account from the Stripe dashboard.
2. Restart your local WCPay server memcache instance to clear cached account data.
3. Refresh account data on merchant store with the dev tools.
4. Under Payment settings, make sure SEPA is enabled.
5. Checkout with a subscription product and repeat steps 5, 6, and 7 from above.

**Note:** While the scenarios above will cover the primary expected areas of impact, it wouldn't hurt to confirm some processes with WooPay enabled and disabled in non-UPE and legacy UPE as well.






-------------------


**Post merge**
